### ISSUE TITLE
Bug: RCS1164: Unused type parameter

### ISSUE THREAD

**Issue by marco-eckstein** [JosefPihrt/Roslynator#1107]
**Product and Version Used**: 

NuGet Package Roslynator.Analyzers 4.3.0

**Steps to Reproduce**:

```csharp
static class Demo
    {
        public static string A() =>
            Run(i => "#" + i); // Calls: Run<T>(Func<int, T>)

        public static void B() =>
            Run<Action>(_ => { }); // Calls: Run<T>(Action<int>)

        static T Run<T>(Func<int, T> block) =>
            block(1);

        static void Run<T>(Action<int> block) =>
            Run(_ => { block(2); return "ignored"; }); // Calls: Run<T>(Func<int, T>)
    }
```

**Actual Behavior**:
```csharp
static void Run<T>(Action<int> block) => // RCS1164: Unused type parameter <T>
```

**Expected Behavior**:
Does not flag T as unused type parameter, because if we remove it, we get this:

```csharp
static class Demo
    {
        public static string A() =>
            Run(i => "#" + i);

        public static void B() =>
            Run<Action>(_ => { }); // CS1643 Not all code paths return a value in lambda expression of type 'Func<int, Action>'


        static T Run<T>(Func<int, T> block) =>
            block(1);

        static void Run(Action<int> block) => // RCS1213: Remove unused member declaration
            Run(_ => { block(2); return "ignored"; });
    }
```

**Comment by JosefPihrt** [comment]
Hi,

I agree that applying of code fix should not cause compiler error.

BUT this analyzer checks if a type parameter is actually used INSIDE the method. I can see the the type parameter is actually not used in the method in which case it makes sense to me to report it and remove it.

**Comment by seruminar** [comment]
Hi @JosefPihrt, how about this example 🙂
```csharp
public static class TestClass
{
    public static void RunDelegate<T>(object arg) // RCS1164: Unused type parameter <T>
    {
        RunDelegate(arg, (T arg) =>
        {
            if (arg is string stringArg)
            {
                Console.WriteLine(stringArg);
            }
        }
        );
    }

    private static void RunDelegate(
        object arg,
        Delegate _delegate
        )
    {
        _delegate.DynamicInvoke(arg);
    }
}
```
```csharp
TestClass.RunDelegate<int>(1)
TestClass.RunDelegate<string>("prints")
prints
```

**Comment by JosefPihrt** [comment]
@seruminar That seems like a bug 👍
